### PR TITLE
Refactoring gen.rs

### PR DESCRIPTION
**Pull request by Yamakaky**
Various refactoring, separated by commits.

---
This change is [Reviewable](https://reviewable.io/reviews/crabtw/rust-bindgen/296)

### CODE REVIEW

**Comment by Yamakaky** [comment]
It should be straightforward to review. Please pay special attention to 02ae5eb57146ba42de85c346a357493df218cf6b and a81afc5812db691aa37fae7a006586ea64c9d2d8.

**Comment by Yamakaky** [comment]
Possibility: move nearly all functions in a `impl GenCtx`.

**Comment by nox** [comment]
So many commits. :) I'll look at it when I go back in Paris. Great work!

**Comment by Yamakaky** [comment]
I tried to separate my commits to facilitate the review :)

File: src/gen.rs [hunk 1]
```diff
@@ -33,8 +33,8 @@ fn ref_eq<T>(thing: &T, other: &T) -> bool {
     (thing as *const T) == (other as *const T)
 }
 
-fn to_intern_str(ctx: &mut GenCtx, s: String) -> parse::token::InternedString {
-    let id = ctx.ext_cx.ident_of(&s[..]);
+fn to_intern_str(ctx: &mut GenCtx, s: &str) -> parse::token::InternedString {
```
Comment (nox) [inline src/gen.rs]: What about just using [`InternedString::new`](http://manishearth.github.io/rust-internals-docs/syntax/parse/token/struct.InternedString.html)?

File: src/gen.rs [hunk 2]
```diff
@@ -38,14 +38,13 @@ fn to_intern_str(ctx: &mut GenCtx, s: &str) -> parse::token::InternedString {
...
 
-fn rust_id(ctx: &mut GenCtx, name: String) -> (String, bool) {
-    let token = parse::token::Ident(ctx.ext_cx.ident_of(&name[..]), parse::token::Plain);
-    if token.is_any_keyword() || "bool" == &name[..] {
-        let mut s = "_".to_owned();
-        s.push_str(&name[..]);
+fn rust_id(ctx: &mut GenCtx, name: &str) -> (String, bool) {
+    let token = parse::token::Ident(ctx.ext_cx.ident_of(name), parse::token::Plain);
+    if token.is_any_keyword() || "bool" == name {
+        let s = format!("_{}", name);
```
Comment (nox) [inline src/gen.rs]: We should actually append an underscore instead of prepend, that's what everyone else in Rust land is doing. Of course we shouldn't do that in that particular commit though, I'm just writing that here so that we don't forget about it.

File: src/gen.rs [hunk 3]
```diff
@@ -168,7 +168,7 @@ pub fn gen_mod(
             GCompDecl(ci) => {
                 {
                     let mut c = ci.borrow_mut();
-                    c.name = unnamed_name(&mut ctx, c.name.clone());
```
Comment (nox) [inline src/gen.rs]: Judging from these line of codes, we are still allocating a new `String` for nothing in most cases.

File: src/gen.rs [hunk 4]
```diff
@@ -155,6 +154,34 @@ fn extract_definitions(ctx: &mut GenCtx,
...
+                        &*sig.ret_ty, &sig.args[..],
+                        sig.is_variadic);
+                        (sig.abi, decl)
+                    }
+                    _ => unreachable!()
+                }
+            },
+            _ => unreachable!()
+        }
+    });
```
Comment (nox) [inline src/gen.rs]: We should have additional methods on `Global` instead of matching, see the first `impl Global` block in `types.rs`.

File: src/gen.rs [hunk 5]
```diff
@@ -274,7 +274,7 @@ pub fn gen_mod(
 }
 
 fn mk_extern(ctx: &mut GenCtx, links: &[(String, LinkType)],
-             foreign_items: Vec<ast::ForeignItem>,
+             foreign_items: &[ast::ForeignItem],
```
Comment (nox) [inline src/gen.rs]: I don't understand this change, we are cloning all the `ForeignItem` values now.

File: src/gen.rs [hunk 6]
```diff
@@ -331,25 +331,20 @@ fn mk_extern(ctx: &mut GenCtx, links: &[(String, LinkType)],
     })
 }
 
-fn remove_redundant_decl(gs: Vec<Global>) -> Vec<Global> {
+fn remove_redundant_decl(gs: &[Global]) -> Vec<Global> {
```
Comment (nox) [inline src/gen.rs]: Same here, why not steal ownership of the `Vec`?

File: src/gen.rs [hunk 7]
```diff
@@ -331,32 +326,6 @@ fn mk_extern(ctx: &mut GenCtx, links: &[(String, LinkType)],
     })
 }
 
-fn remove_redundant_decl(gs: &[Global]) -> Vec<Global> {
```
Comment (nox) [inline src/gen.rs]: This probably indicates a lack of tests, not that this code is useless.

File: src/gen.rs [hunk 8]
```diff
@@ -326,57 +325,6 @@ fn mk_extern(ctx: &mut GenCtx, links: &[(String, LinkType)],
     })
 }
 
-fn tag_dup_decl(gs: &[Global]) -> Vec<Global> {
```
Comment (nox) [inline src/gen.rs]: Same, either we lack tests or removing duplicates is broken, in both cases this code probably is supposed to be useless.

File: src/gen.rs [hunk 9]
```diff
@@ -589,25 +589,31 @@ fn cunion_to_rs(ctx: &mut GenCtx, name: String, derive_debug: bool, layout: Layo
...
-    let int_lit = ast::LitKind::Int(
-        val.abs() as u64,
-        ast::LitIntType::Unsuffixed
-            );
-    let mut value = ctx.ext_cx.expr_lit(ctx.span, int_lit);
-    if val < 0 {
-        let negated = ast::ExprKind::Unary(ast::UnOp::Neg, value);
-        value = ctx.ext_cx.expr(ctx.span, negated);
+/// Converts a signed number to AST Expression.
+fn i64_to_int_lit(ctx: &mut GenCtx, value: i64) -> P<ast::Expr> {
```
Comment (nox) [inline src/gen.rs]: I'm pretty sure this is strictly the same function as `cenum_value_to_int_lit`.

File: src/gen.rs [hunk 10]
```diff
@@ -653,15 +654,7 @@ fn cenum_value_to_int_lit(
...
-        let int_lit =
-            ast::LitKind::Int(value.abs() as u64, ast::LitIntType::Unsuffixed);
-        let expr = ctx.ext_cx.expr_lit(ctx.span, int_lit);
-        if value < 0 {
-            ctx.ext_cx.expr(
-                ctx.span, ast::ExprKind::Unary(ast::UnOp::Neg, expr))
-        } else {
-            expr
-        }
+        i64_to_int_lit(ctx, value)
```
Comment (nox) [inline src/gen.rs]: `i64_to_int_lit` should actually handle sign too.

File: src/gen.rs [hunk 11]
```diff
@@ -1128,17 +1128,14 @@ fn mk_ty(ctx: &GenCtx, global: bool, segments: Vec<String>) -> ast::Ty {
     }
 }
 
-fn mk_ptrty(ctx: &mut GenCtx, base: &ast::Ty, is_const: bool) -> ast::Ty {
-    let ty = ast::TyKind::Ptr(ast::MutTy {
-        ty: P(base.clone()),
-        mutbl: if is_const { ast::Mutability::Immutable } else { ast::Mutability::Mutable }
-    });
+fn mk_ptrty(ctx: &mut GenCtx, base: ast::Ty, is_const: bool) -> ast::Ty {
```
Comment (nox) [inline src/gen.rs]: Why not pass `P<Ty>` actually?

File: src/gen.rs (hunk 1 above)
Comment (Yamakaky) [inline src/gen.rs]: Seems good ^^

File: src/gen.rs (hunk 2 above)
Comment (Yamakaky) [inline src/gen.rs]: Maybe open an issue, your comment will be lost ^^

File: src/gen.rs (hunk 3 above)
Comment (Yamakaky) [inline src/gen.rs]: ? why ?

File: src/gen.rs (hunk 5 above)
Comment (Yamakaky) [inline src/gen.rs]: no, Vec<ForeignItem> is Deref<[ForeignItem]>

Comment (nox) [inline src/gen.rs]: Sure, but you are cloning the items in the function by doing `foreign_items.iter().cloned()`.

File: src/gen.rs (hunk 3 above)
Comment (nox) [inline src/gen.rs]: In most cases, you just want to return the name as is. So we should just make unnamed_name steal `c.name` and put it back inside.

Comment (nox) [inline src/gen.rs]: We could otherwise make it take `&mut String`.

File: src/gen.rs (hunk 5 above)
Comment (Yamakaky) [inline src/gen.rs]: Oh, yeah. In fact, it was to make this function more general, but maybe I have been too far ^^

**Comment by nox** [comment]
@Yamakaky Ping?

**Comment by Yamakaky** [comment]
I was on another project, I need to go back to bindgen ;)

File: src/gen.rs (hunk 1 above)
Comment (Yamakaky) [inline src/gen.rs]: In fact, no: InternedString::new takes a &'static str, while to_intern_str takes a &str.

File: src/gen.rs (hunk 4 above)
Comment (Yamakaky) [inline src/gen.rs]: I plan on doing it after this PR is merged, it's already too big ^^

File: src/gen.rs (hunk 1 above)
Comment (nox) [inline src/gen.rs]: Oh right. Then just `InternedString::new` where we pass a literal string and `ctx.ext_cx.name_of(s).as_str()` everywhere else, no need for an intermediate function or `Ident` value.

File: src/gen.rs (hunk 4 above)
Comment (nox) [inline src/gen.rs]: Ok.

File: src/gen.rs (hunk 5 above)
Comment (Yamakaky) [inline src/gen.rs]: In fact, this function is only used gen_mod with Vecs, so maybe I should just revert the change?

Comment (nox) [inline src/gen.rs]: Sounds good.

File: src/gen.rs (hunk 8 above)
Comment (Yamakaky) [inline src/gen.rs]: Reverting for now. #305

File: src/gen.rs (hunk 9 above)
Comment (Yamakaky) [inline src/gen.rs]: You are right, except for the `& enum_size_to_unsigned_max_value(size);` in cenum_value_to_int_lit. In fact, it's strange because it only does this for unsigned enum variants. Maybe a bug?

File: src/gen.rs (hunk 11 above)
Comment (Yamakaky) [inline src/gen.rs]: Because cty_to_rs returns a Ty. See in cty_to_rs.

File: src/gen.rs (hunk 9 above)
Comment (nox) [inline src/gen.rs]: No, overflow and wrapping of signed values is undefined behaviour in C, so we don't and shouldn't handle them. Just let it emit an error at compile-time on Rust side.

File: src/gen.rs (hunk 11 above)
Comment (nox) [inline src/gen.rs]: Yeah I know that, but `cty_to_rs` should return `P<Ty>` too.

Comment (Yamakaky) [inline src/gen.rs]: I don't really see why P exists. Why not using Box?

Comment (nox) [inline src/gen.rs]: It's a refcounted pointer. Box is ownership.

File: src/gen.rs (hunk 9 above)
Comment (Yamakaky) [inline src/gen.rs]: So what should I do?

Comment (nox) [inline src/gen.rs]: Same as for enum values, don't wrap nor truncate signed constants.

File: src/gen.rs (hunk 11 above)
Comment (nox) [inline src/gen.rs]: Actually, [brainfart](http://manishearth.github.io/rust-internals-docs/src/syntax/src/libsyntax/ptr.rs.html#48-50). I guess P just predates Box, given rustc is the oldest Rust application out there.

File: src/gen.rs (hunk 9 above)
Comment (Yamakaky) [inline src/gen.rs]: So without `& enum_size_to_unsigned_max_value(size);` for the two functions?

Comment (nox) [inline src/gen.rs]: Just use `cenum_value_to_int_lit` in both places, no? Maybe rename it.

File: src/gen.rs (hunk 11 above)
Comment (Yamakaky) [inline src/gen.rs]: That's what I thought.

File: src/gen.rs (hunk 9 above)
Comment (Yamakaky) [inline src/gen.rs]: Ok

Comment (Yamakaky) [inline src/gen.rs]: Not sure if you saw it, but i64_to_int_lit is also used in const_to_rs. I'm not sure the semantics of cenum_value_to_int_lit matches.

Comment (nox) [inline src/gen.rs]: AFAICT, we want the same behaviour for const than we do for enum values. That's just the same thing about coercing a literal to a given integer type.

Cf our discussion [there](https://github.com/crabtw/rust-bindgen/pull/263#issuecomment-192652267).

File: src/gen.rs (hunk 10 above)
Comment (Yamakaky) [inline src/gen.rs]: It does.

**Comment by Yamakaky** [comment]
Should have corrected the issues, ping me if I forgot something. I'll stop adding more commits to this PR if possible XD

**Comment by Yamakaky** [comment]
The build failure is about apt:
```
E: Unable to locate package libclang-3.8-dev

E: Couldn't find any package by regex 'libclang-3.8-dev'

E: Unable to locate package llvm-3.8-dev

E: Couldn't find any package by regex 'llvm-3.8-dev'
```

**Comment by Yamakaky** [comment]
@nox I think this PR is ready for merge?

**Comment by nox** [comment]
SGTM.
